be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa *)));
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa)));
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(struct m0_dtm0_log_rec));
+}
+
+M0_INTERNAL void dtm0_be_del_log_rec_credit(struct m0_be_seg       *seg,
+					    struct m0_dtm0_log_rec *rec,
+					    struct m0_be_tx_credit *accum)
+{
+	lrec_be_list_credit(M0_BLO_TLINK_DESTROY, 1, accum);
+	M0_BE_FREE_CREDIT_ARR(rec->dlr_txd.dtd_ps.dtp_pa,
+			      rec->dlr_txd.dtd_ps.dtp_nr, seg, accum);
+	M0_BE_FREE_CREDIT_PTR(rec->dlr_payload.b_addr, seg, accum);
+	M0_BE_FREE_CREDIT_PTR(rec, seg, accum);
+}
+
+M0_INTERNAL void dtm0_be_log_destroy_credit(struct m0_be_seg       *seg,
+					    struct m0_be_tx_credit *accum)
+{
+	lrec_be_list_credit(M0_BLO_DESTROY, 1, accum);
+}
+
+M0_INTERNAL void m0_be_dtm0_log_credit(enum m0_be_dtm0_log_credit_op op,
+				       struct m0_dtm0_tx_desc	    *txd,
+				       struct m0_buf		    *payload,
+				       struct m0_be_seg             *seg,
+				       struct m0_dtm0_log_rec	    *rec,
+				       struct m0_be_tx_credit       *accum)
+{
+	/*TODO:Complete implementation during persistent list implementation */
+	struct m0_be_dtm0_log *log;
+
+	switch (op) {
+	case M0_DTML_CREATE:
+		M0_BE_ALLOC_CREDIT_PTR(log, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->dl_seg, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->u.dl_list, seg, accum);
+		lrec_be_list_credit(M0_BLO_CREATE, 1, accum);
+		m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_PTR(log));
+		break;
+	case M0_DTML_EXECUTED:
+	case M0_DTML_PERSISTENT:
+		dtm0_be_alloc_log_rec_credit(txd, payload, seg, accum);
+		dtm0_be_set_log_rec_credit(txd, payload, seg, accum);
+		m0_be_list_credit(M0_BLO_ADD, 1, accum);
+		break;
+	case M0_DTML_PRUNE:
+		dtm0_be_del_log_rec_credit(seg, rec, accum);
+		m0_be_list_credit(M0_BLO_DEL, 1, accum);
+		break;
+	case M0_DTML_REDO:
+	default:
+		M0_IMPOSSIBLE("");
+	}
+}
+
+M0_INTERNAL int m0_be_dtm0_log_create(struct m0_be_tx        *tx,
+                                      struct m0_be_seg       *seg,
+                                      struct m0_be_dtm0_log **out)
+{
+	struct m0_be_dtm0_log *log;
+
+	M0_PRE(tx != NULL);
+	M0_PRE(seg != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log, seg, tx);
+	M0_ASSERT(log != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log->u.dl_list, seg, tx);
+	M0_ASSERT(log->u.dl_list != NULL);
+
+	log->dl_seg = seg;
+
+	lrec_be_list_create(log->u.dl_list, tx);
+	M0_BE_TX_CAPTURE_PTR(seg, tx, log);
+	*out = log;
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_destroy(struct m0_be_tx        *tx,
+                                        struct m0_be_dtm0_log **log)
+{
+#if 0 /* TODO: write down persistent implementation later */
+	struct m0_be_dtm0_log *plog = *log;
+
+	M0_PRE(plog != NULL);
+	m0_free(plog->u.dl_tlist);
+	m0_free(plog);
+	*log = NULL;
```

Comment:
This piece of code is misleading and wrong.
For a persistent log, we can not use m0_free().
Also we can only use "dl_list" here.

Reply:
I have removed it.